### PR TITLE
Add Prometheus support

### PR DESCRIPTION
**Pull request by kereis**
See #193

### CODE REVIEW

File: prometheus_metrics/measurements.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+// or to gaugeVecMap - Otherwise Prometheus won't keep track of the newly added meters.Measurement
+var counterVecMap = map[meters.Measurement]*prometheus.CounterVec{}
+
+// gaugeVecMap contains all meters.Measurement that are associated with a prometheus.Gauge
+//
+// If a new meters.Measurement is introduced, it needs to be added either to counterVecMap
+// or to gaugeVecMap - Otherwise Prometheus won't keep track of the newly added meters.Measurement
+var gaugeVecMap = map[meters.Measurement]*prometheus.GaugeVec{}
+
+// UpdateMeasurementMetric updates a counter or gauge based by passed measurement
```
Comment (ngrsdhbw) [inline prometheus_metrics/measurements.go:55]: based by-->based on the

**Review by ngrsdhbw — CHANGES_REQUESTED** [review]

File: meters/units.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+type unit struct {
+	Abbreviation unitAbbreviation
+	FullName     string
+	PluralForm   string
+}
+
+// unitAbbreviation defines the default abbreviation and - if needed - an alternative.
+type unitAbbreviation struct {
+	Default     string
+	Alternative string
```
[the remark this answers is not in the document]
Comment (kereis) [reply meters/units.go:40]: If you use `°` for degree, you cannot use the special character in the Prometheus metrics name. That's why `Alternative` will (always) be used instead if it is defined.

[the remark this answers is not in the document]
Comment (ngrsdhbw) [reply meters/units.go:40]: then  it should be named metrics_alternative or something like that

File: meters/units.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+var units = map[Unit]*unit{
+	KiloVarHour:   {unitAbbreviation{"kvarh", ""}, "Kilovoltampere-hours (reactive)", ""},
+	Var:           {unitAbbreviation{"var", "volt_ampere"}, "Voltamperes (reactive)", "volt_amperes"},
+	KiloWattHour:  {unitAbbreviation{"kWh", ""}, "Kilowatt-hours", ""},
+	Watt:          {unitAbbreviation{"W", ""}, "Watts", ""},
+	Ampere:        {unitAbbreviation{"A", ""}, "Amperes", ""},
+	Volt:          {unitAbbreviation{"V", ""}, "Volts", ""},
+	VoltAmpere:    {unitAbbreviation{"VA", ""}, "Voltamperes", ""},
```
[the remark this answers is not in the document]
Comment (kereis) [reply meters/units.go:50]: I think I have to refactor this one.

If the `PluralForm` is not defined, then we simply take the enum's "name" as string and append an `s` (as the units' names are in singular). The syntax of the name of one enum's element is defined in the `go:generate` statement for enumer. I've set it to snake-case.

Previously, I've used the snake-case format in `PluralForm` for Prometheus names. However, if we want to use the plural form in the help texts as well, we have to set the actual `FullName`.

File: meters/units.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	Hertz:         {unitAbbreviation{"Hz", "hertz"}, "Hertz", "hertz"},
+	Percent:       {unitAbbreviation{"%", "percent"}, "Per cent", "percent"},
+	Joule:         {unitAbbreviation{"J", ""}, "Joules", ""},
+
+	NoUnit: {unitAbbreviation{"", ""}, "", ""},
+}
+
+var conversionMap = map[Unit]func(float64) float64{
+	Joule: func(input float64) float64 {
+		return (input * 3.6) * 1_000_000
```
[the remark this answers is not in the document]
Comment (kereis) [reply meters/units.go:62]: For this, I went with the "formula":

```
J = (x kWh * 3.6 MJ) * 1 000 000
```

1. Convert `kWh` to `MJ`
2. Convert `MJ` to `J`

[the remark this answers is not in the document]
Comment (ngrsdhbw) [reply meters/units.go:62]: doesn´t the compiler optimize this anyway?
in my opinion first removing the metric prefix and then converting the time base is more intuitive

File: meters/units.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kereis) [reply meters/units.go:50]: We could think about if we actually need the singular form and just rename enum elements itself - which may render `PluralForm` as redundant.

File: meters/measurements.go [hunk 5]
```diff
@@ -142,99 +145,99 @@ const (
...
+	THDL2:            newInternalMeasurement(withDescription("L2 Voltage to neutral THD"), withUnit(units.Percent), withMetricType(Gauge)),
+	THDL3:            newInternalMeasurement(withDescription("L3 Voltage to neutral THD"), withUnit(units.Percent), withMetricType(Gauge)),
+	Sum:              newInternalMeasurement(withDescription("Total Energy Sum"), withPrometheusName("energy_sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
+	SumT1:            newInternalMeasurement(withDescription("Tariff 1 Energy Sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
+	SumT2:            newInternalMeasurement(withDescription("Tariff 2 Energy Sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
+	SumL1:            newInternalMeasurement(withDescription("L1 Energy Sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
+	SumL2:            newInternalMeasurement(withDescription("L2 Energy Sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
+	SumL3:            newInternalMeasurement(withDescription("L3 Energy Sum"), withUnit(units.KiloWattHour), withMetricType(Counter)),
```
Comment (kereis) [inline meters/measurements.go:191]: @andig What does `Total [Energy] Sum` exactly mean? Does it represent a kind of balance between imported/exported energy?

Comment (RichieB2B) [reply meters/measurements.go:191]: I think this is just the `Sum` value that `mbmd` is reading from the meters. Whether this is actually a sum of import and export (probably import - export) is up to the meter manufacturer.

**Comment by SchumacherFM** [comment]
any update on this?

**Comment by RichieB2B** [comment]
While not the same as a native Prometheus exporter, I am now using a python script to get the mbmd data via mqtt into Prometheus: https://github.com/RichieB2B/simple-mqtt-exporter/blob/main/examples/config-mbmd.py

**Comment by kereis** [comment]
Hey, I'm sorry, but I'm afraid I cannot provide real support or updates on this one. This implementation has been part of a research project some years before, and I can't guarantee its functionality anymore as e.g. I have no way to test this in a local environment properly.

Feel free to pick up on this MR if you think that the current code base suits your needs.

**Comment by SchumacherFM** [comment]
@kereis Thanks for the update! Once my SolarEdge inverters are running will continue with your PR/MR.

**Pull request closed without merging**